alluxio-ui/master/src/containers/pages/Browse/Browse.tsx

```diff
@@ -0,0 +1,277 @@
+      browse.invalidPathError || browse.invalidPathException) {
+      return (
+        <Alert color="danger">
+          {errors && <div>Unable to reach the api endpoint for this page.</div>}
+          {browse.accessControlException && <div>{browse.accessControlException}</div>}
+          {browse.fatalError && <div>{browse.fatalError}</div>}
+          {browse.fileDoesNotExistException && <div>{browse.fileDoesNotExistException}</div>}
+          {browse.invalidPathError && <div>{browse.invalidPathError}</div>}
+          {browse.invalidPathException && <div>{browse.invalidPathException}</div>}
+        </Alert>
+      );
+    }
+
+    return (
+      <div className="browse-page">
+        <div className="container-fluid">
+          <div className="row">
+            <div className="col-12">
+              {!browse.currentDirectory.isDirectory && this.renderFileView(browse, queryStringSuffix)}
+              {browse.currentDirectory.isDirectory && this.renderDirectoryListing(browse, queryStringSuffix)}
+            </div>
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  private renderFileView(browse: IBrowse, queryStringSuffix: string) {
+    const {textAreaHeight, path, offset, end, lastFetched} = this.state;
+    const offsetInputHandler = this.createInputHandler('offset', value => value).bind(this);
+    const beginInputHandler = this.createInputHandler('end', value => undefined).bind(this);
+    const endInputHandler = this.createInputHandler('end', value => '1').bind(this);
+    return (
+      <React.Fragment>
+        <FileView allowDownload={true} beginInputHandler={beginInputHandler} end={end} endInputHandler={endInputHandler}
+                  lastFetched={lastFetched} offset={offset} offsetInputHandler={offsetInputHandler} path={path}
+                  queryStringPrefix="/browse" queryStringSuffix={queryStringSuffix} textAreaHeight={textAreaHeight}
+                  viewData={browse}/>
+        <hr/>
+        <h6>Detailed blocks information (block capacity is {browse.blockSizeBytes} Bytes):</h6>
+        <Table hover={true}>
+          <thead>
+          <tr>
+            <th>ID</th>
+            <th>Size (Byte)</th>
+            <th>In {browse.highestTierAlias}</th>
+            <th>Locations</th>
+          </tr>
+          </thead>
+          <tbody>
+          {browse.fileBlocks.map((fileBlock: IFileBlockInfo) => (
+            <tr key={fileBlock.id}>
+              <td>{fileBlock.id}</td>
+              <td>{fileBlock.blockLength}</td>
+              <td>
+                {/*TODO: fix this to be part of the fileBlock in the api*/}
+                {browse.highestTierAlias ? 'YES' : 'NO'}
+              </td>
+              <td>
+                {fileBlock.locations.map((location: string) => (
+                  <div key={location}>{location}</div>
+                ))}
+              </td>
+            </tr>
+          ))}
+          </tbody>
+        </Table>
+      </React.Fragment>
```

Comment:
A TODO for future refactoring, is there anyway we can make this more modular as opposed to having this method spit out the entire dom object?

Reply:
absolutely, it can be split up into different sub-components